### ISSUE TITLE
Allow onRoute to alter options without side effects

### ISSUE THREAD

**Issue by mcollina** [fastify/fastify#2767]
## 🚀 Feature Proposal

The way the `onRoute` hook can be used is to derive some info from the route settings (for swagger for example), or to __mutate the route definition itself__. However in certain cases Fastify creates two routes from the same definition  (see https://github.com/fastify/fastify/blob/834a4e58de86d7996b50a4663fbf2b0fac4686ee/lib/route.js#L163-L170).

I propose to update the `onRoute` hook so that:

```js
fastify.addHook('onRoute', function (opts) {
  const copy = Object.assign({}, opts)
  copy.onRequest = newHook
  return copy
})
```

This is to avoid having some custom state like in: https://github.com/fastify/fastify-compress/pull/141.

Note that the actual change is quite simple, as we just need to alter:
https://github.com/fastify/fastify/blob/834a4e58de86d7996b50a4663fbf2b0fac4686ee/lib/route.js#L200-L207.

This change is semver-major and should target the `next`  branch.

**Comment by mcollina** [comment]
This is another example of this problem: https://github.com/fastify/fastify-rate-limit/issues/120.

**Comment by jsumners** [comment]
👍

**Comment by MomenNano** [comment]
can we just remove the second `afterRouteAdded.call`  here in line 168 
🤔

https://github.com/fastify/fastify/blob/834a4e58de86d7996b50a4663fbf2b0fac4686ee/lib/route.js#L163-L170

**Comment by mcollina** [comment]
> can we just remove the second afterRouteAdded.call here in line 168

Unfortunately no. That's needed to support the `ignoreTrailingSlash` option.

Moreover, it would still be a bug, consider the following case:

```js
const opts = {
  onRequest: [...]
}
fastify.post('/', opts, handler)
fastify.put('/', opts, handler)
```

You'd see the exact same problem.

**Comment by AyoubElk** [comment]
I'd like to work on this @mcollina, can you clarify the desired behavior for me?

**Comment by mcollina** [comment]
> I'd like to work on this @mcollina, can you clarify the desired behavior for me?
> 
> 
> 
> You mentioned that a copy of the route options is created using `Object.assign`, but what happens next?
> 
> Should that copy be used to create the context and register the route on the router? In this case we'd be trying to register X new routes from the same original route definition (where X is the number of `onRoute` hooks executed, which btw wouldn't have necessarily mutated the route definition)

No. Let's think of `onRoute` as a `reduce` operation: we register only the last.

**Comment by AyoubElk** [comment]
> > I'd like to work on this @mcollina, can you clarify the desired behavior for me?
> > You mentioned that a copy of the route options is created using `Object.assign`, but what happens next?
> > Should that copy be used to create the context and register the route on the router? In this case we'd be trying to register X new routes from the same original route definition (where X is the number of `onRoute` hooks executed, which btw wouldn't have necessarily mutated the route definition)
> 
> No. Let's think of `onRoute` as a `reduce` operation: we register only the last.

Got it! 

Do you think a deep copy is more appropriate in this context instead of `Object.assign()`?

**Comment by mcollina** [comment]
> Do you think a deep copy is more appropriate in this context instead of Object.assign()?

I would leave this to the implementor of the hook. It might just be 2 levels (main object and hooks), or a partial tree.

**Comment by simoneb** [comment]
I'm having a look at this and wondering if there isn't an existing fundamental issue in the handling of the trailing slash.

https://github.com/fastify/fastify/blob/834a4e58de86d7996b50a4663fbf2b0fac4686ee/lib/route.js#L163-L170

If the conditional is entered, we have two async operations calling the same callback, right? I understand it may not in practice lead to big issues, but it doesn't look right either.

**Comment by RafaelGSS** [comment]
@simoneb Actually there will have two routes: `/path` and `/path/`.

**Comment by simoneb** [comment]
> @simoneb Actually there will have two routes: `/path` and `/path/`.

Yes I can see that, but both are calling the same `done` callback.

**Comment by mcollina** [comment]
> I'm having a look at this and wondering if there isn't an existing fundamental issue in the handling of the trailing slash.
> 
> 
> 
> https://github.com/fastify/fastify/blob/834a4e58de86d7996b50a4663fbf2b0fac4686ee/lib/route.js#L163-L170
> 
> 
> 
> If the conditional is entered, we have two async operations calling the same callback, right? I understand it may not in practice lead to big issues, but it doesn't look right either.

That's correct and expected. That's how we support the trailing slash.

**Comment by simoneb** [comment]
I guess I'm missing something then, I don't understand how calling the `done` callback twice is correct. I tried tracing back to where that callback comes from and it's buried somewhere in the avvio package so I lost track of how it's handled. Conceptually, that callback is used to communicate to the caller that the async operation is completed and the caller may be waiting for that before continuing doing something else, so calling it twice doesn't seem right to me but I appreciate I may be missing context.

**Comment by jsumners** [comment]
@simoneb only one route will be hit for a request. So the `done` is only invoked once per route handled. `'/foo' == '/foo/'` in terms the handler.

**Comment by simoneb** [comment]
It looks like I'm not explaining myself. This is not something that happens when the routes are hit, but when they are registered.

https://github.com/fastify/fastify/blob/834a4e58de86d7996b50a4663fbf2b0fac4686ee/lib/route.js#L163-L170

This causes the `done` callback to be called twice. I traced it back to [this function in avvio](https://github.com/fastify/avvio/blob/24c4ff1d163f8264d8714dfe9ca6d0b215cbd009/boot.js#L452-L462).

You can see that in the else branch nothing's done. If that function was modified to set the timer to null when it's cleared and throw an exception if timer is null, then under the conditions of this issue (two routes registered), you'll get an exception because the `done` callback is called twice.

Now this may not be a big issue, that callback is notifying avvio that the plugin has finished registering, but in this case it's signalling that it's finished registering after the first route is registered, while the second still has to be registered. I don't know what kind of consequences this may have, but I guess some code waiting for the full registration of the plugin may not work correctly.

I appreciate this may be a kind of extreme edge case though.

**Comment by salmanm** [comment]
I'd like to present my take on the original issue.

The way we prevented acting twice in https://github.com/fastify/fastify-compress/pull/141 was by adding a property to the `routeOptions`. However, if we somehow knew in `onRoute` if it's being called for the same route as before but just for prefixing a slash then we'd be able to avoid adding custom property right?

How do you guys feel if we get a `routeOptions.prefixing` property?

```js
fastify.addHook('onRoute', (routeOptions) => {
  console.log(routeOptions.prefixing) // true when relevant
})
```

With this `fastify-compress` could detect when not to act.

Since it was a simple enough change I [pushed a commit](https://github.com/fastify/fastify/commit/8f9ff98d1a4a3d978ac3965bbc7e066831f518e7#diff-9d81ec3a6358f6b576e04e6122b7391af9aa3c85d6491e640c78a89c58414c4fR565-R592) to demonstrate.

Let me know what you think.

**Comment by RafaelGSS** [comment]
I might be wrong, but for me, the hook is working properly since it adds two routes when `ignoreTrailingSlash = true`. But yeah, the use case is quite complex.

I would change to a different approach, instead of adding a property to identify if the route has the same handler as others. How complex should be register the route only once but on lookup we transform `/path/` to `/path`? I mean, this is only an idea to fix the problem above, not sure if it's related to the current issue, I would create a new issue for that though.

**Comment by mcollina** [comment]
> I would change to a different approach, instead of adding a property to identify if the route has the same handler as others. How complex should be register the route only once but on lookup we transform /path/ to /path

This will slow us down :(
